File: 2023/go/day1/1.go
package day1

import (
	"fmt"
	"regexp"
	"strconv"
	"strings"
	"unicode"

	log "github.com/sirupsen/logrus"
)

func Puzzle(input *[]byte, part2 bool) int {
	//log.SetLevel(log.DebugLevel)
	total := 0
	for _, line := range strings.Split(strings.TrimSpace(string(*input)), "\n") {
		log.Debug(line)
		calibrationString := line
		if part2 {
			calibrationString = parseDigit(line)
		}
		re := regexp.MustCompile("[0-9]")
		digits := re.FindAllString(calibrationString, -1)
		log.Debug(digits)
		var first, last string
		for _, digit := range digits {
			if first == "" {
				first = digit
			}
			last = digit
		}
		valueStr := first + last
		log.Debug(valueStr)
		calibrationValue, err := strconv.Atoi(valueStr)
		if err != nil {
			log.Error(err)
		}
		total += calibrationValue
	}
	return total
}

func parseDigit(cs string) string {
	digits := []string{"one", "two", "three", "four", "five", "six", "seven", "eight", "nine"}
	ncs := ""
	for i, r := range cs {
		if unicode.IsDigit(r) {
			ncs += string(r)
			continue
		}
		for d, digit := range digits {
			re := regexp.MustCompile(digit)
			if re.Match([]byte(cs[i:])) {
				ncs += fmt.Sprint(d + 1)
			}
		}
	}
	return ncs
}

File: 2023/python/1.py
#!/usr/bin/env python3

import re
import time
startTime = time.time()

def puzzle(input, p2):
    numbers = [] 
    for line in input.splitlines():
        if p2:
            line = parseDigits(line)
        digits = re.findall(r'\d', line)
        numbers.append(int(digits[0] + digits[len(digits)-1]))
    return sum(numbers)

def parseDigits(line):
    digits = [ r"one", r"two", r"three", r"four", r"five", r"six", r"seven", r"eight", r"nine" ]
    calibrationStr = ""
    for i, c in enumerate(line):
        if c.isdigit():
            calibrationStr += c
            continue
        for j, digit in enumerate(digits):
            match = re.match(digit, line[i:])
            if match: 
                calibrationStr += str(j+1)
    return calibrationStr

f = open("2023/input/1", "r")
input = f.read()

print("2023 - Day 1 - Part 1:", puzzle(input, False))
print("2023 - Day 1 - Part 2:", puzzle(input, True))
print("Execution time: %.3fms" % ((time.time() - startTime)*1000))